feature: MongoDB support via beanie

usefulalgorithm opened the pull request:
Adds support for MongoDB and Beanie. Probably fixes #119 ?

I did a simple crud app with virtually the same setup, seems to be working fine.

## Demos
- project setup: https://asciinema.org/a/U5YaCeweA4vGqmNCv5zpBpwSc
- deploying MongoDB and starting the application: https://asciinema.org/a/XOq7NByeER8kC3tT0cLFsW2gO
- a simple crud app: https://github.com/usefulalgorithm/fastapi-mongo-demo-app/
  - some differences between this project and this PR:
    - `db` service is placed in `docker-compose.dev.yaml` instead of `docker-compose.yml`
    - `Settings` has a separate section for mongodb settings

## Todos / caveats / things to note:
- MongoDB requires additional setup if you're not writing to the `admin` database
- Some docstring and comments are probably off, and flake is likely to complain.
- I have no idea how cookiecutter works, so I basically did it like how I use Jinja templates and tried to fill in as many gaps as possible. I'm guessing I probably missed a lot of places...
- The only setup I tested is the one in the clip.

s3rius commented on fastapi_template/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/web/lifetime.py:
```diff
@@ -120,6 +120,24 @@ def _setup_db(app: FastAPI) -> None:  # pragma: no cover
+from {{cookiecutter.project_name}}.db.models.dummy_model import DummyModel
+async def _setup_db(app: FastAPI) -> None:
+    client = AsyncIOMotorClient(
+        f"mongodb://{settings.db_user}:{settings.db_pass}"
```
Please use yarl to build URLs.

Take a look at `db_url` property in settings.

usefulalgorithm replied:
ok! thanks for the tips

s3rius commented on fastapi_template/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/web/api/dummy/views.py:
```diff
@@ -14,28 +16,44 @@
+{%- if cookiecutter.db_info.name != "mongodb" %}
     await dummy_dao.create_dummy_model(name=new_dummy_object.name)
+{%- else %}
+    await DummyModel.insert_one(DummyModel(name=new_dummy_object.name))
```
Can we move this into DAO? All databases follow this pattern. Is it a good idea for mongo as well?

usefulalgorithm replied:
sure.

s3rius reviewed (changes requested):
Hi and thanks for your AWESOME WORK! But I have some questions and suggestions regarding the PR. 

Few things to add:
* Services for Github actions and Gitlab-Ci. It's used by generate CI option.
* Fix dummy for GraphQL in web/gql folder. Since it's the same as for REST right now.
* Can you please add test in `fastapi_template/tests/test_generator.py`? Because we need to be sure that after the project generation, all generated tests are passing. 

Thank you very much. I really appreciate the work you did.

s3rius commented:
Actually, you can easily test the project generations by creating one command and running it everytime you make changes. For example.

```bash
fastapi_template --name myproject --db mongo --api-type rest --orm beanie --dummy --routers --quite --force
```

This command will be overriding all files within myproject folder. It helps me fixing all flake errors.

s3rius commented:
Also, you should not start DB separately from the project. The main point is that the `up` command will bring all services app, including your application.

```python
docker compose up --build
```

usefulalgorithm commented on pyproject.toml:
```diff
@@ -27,6 +27,7 @@ pytest-env = "^0.6.2"
 Faker = "^8.14.0"
 pytest-xdist = {version = "^2.5.0", extras = ["psutil"]}
 requests = "^2.28.1"
+pynvim = "^0.4.3"
```
probably not... I'll remove them

usefulalgorithm commented on fastapi_template/template/{{cookiecutter.project_name}}/conditional_files.json:
```diff
@@ -56,13 +56,24 @@
         ]
     },
+    "SQLAlchemy support": {
+        "enabled": "{{cookiecutter.db_info.name not in ['none', 'mongodb']}}",
```
I removed a line from the `Database support` dict and made this condition with it.

I actually have no idea what these conditions are for..

usefulalgorithm commented:
Hi @s3rius , I think I addressed all the things you suggested:

- Services for Github actions and Gitlab-Ci. It's used by generate CI option.
  - Added variables in `test.yml` and `.gitlab-ci.yml`.
- Fix dummy for GraphQL in web/gql folder. Since it's the same as for REST right now.
  - Done.
- Can you please add test in fastapi_template/tests/test_generator.py?
  - Added several tests in that file.

I also made sure all flake8 warnings are gone, mypy does not complain, and that I was able to create & read dummy models with the containers started by docker compose.

However I wasn't able to get pytest to run successfully (both in `fastapi_template` and in the generated `myproject`). In `fastapi_template` running `poetry run pytest` breaks:
```
FAILED fastapi_template/tests/test_generator.py::test_default_with_db[sqlalchemy-mysql] - AssertionError
FAILED fastapi_template/tests/test_generator.py::test_default_with_db[tortoise-mysql] - AssertionError
FAILED fastapi_template/tests/test_generator.py::test_default_with_db[ormar-mysql] - AssertionError
FAILED fastapi_template/tests/test_generator.py::test_default_with_nosql_db[beanie-mongodb] - AssertionError
```
and running tests in `myproject` would always end up with an unauthenticated client. Are the tests meant to be run only when triggered by github / gitlab pipelines?


Thanks!

usefulalgorithm commented:
Hi @s3rius, I modified the things you addressed, please check out the changes and see if they make sense when you have time, thanks!

usefulalgorithm commented:
Hi @s3rius , sorry for the lack of response. I've addressed all your suggestions in the latest commit, please take a look when you have time.

There's something I don't get in the generated `test_dummy.py` file:

```python
async def test_creation(
    fastapi_app: FastAPI,
    client: AsyncClient,
) -> None:
    """Tests dummy instance creation."""
    url = fastapi_app.url_path_for("create_dummy_model")
    test_name = uuid.uuid4().hex
    response = await client.put(
        url,
        json={
            "name": test_name,
        },
    )   # Inserts a Dummy into DB
# ... snipped ...
@pytest.mark.anyio
async def test_getting(
    fastapi_app: FastAPI,
    client: AsyncClient,
) -> None:
    """Tests dummy instance retrieval."""
    dao = DummyDAO()
    test_name = uuid.uuid4().hex
    await dao.create_dummy_model(name=test_name)  # XXX Inserts another Dummy into DB??
    url = fastapi_app.url_path_for("get_dummy_models")
    response = await client.get(url)
    dummies = response.json()

    assert response.status_code == status.HTTP_200_OK
    assert len(dummies) == 1   # XXX Should be 2?
```

This the only test that's failing. What I don't understand is that how come in `test_getting` it is testing that the object we inserted is indeed the only object in the database, while the object inserted in `test_creation` is never cleaned up?

Thanks in advance for taking a look and answering the question!

usefulalgorithm commented:
Hi @s3rius , I believe this is complete now. Thanks!

usefulalgorithm commented:
@s3rius  ping

s3rius commented on fastapi_template/template/{{cookiecutter.project_name}}/conditional_files.json:
```diff
@@ -56,13 +56,18 @@
             "alembic.ini",
             "{{cookiecutter.project_name}}/web/api/dummy",
             "{{cookiecutter.project_name}}/web/gql/dummy",
-            "{{cookiecutter.project_name}}/db_sa",
```
Please place this resource back. It will leave db_sa folder in a proejct, if database is disabled.

s3rius replied:
Or maybe not, because of SQLAlchemy. Selector. I will check it.

usefulalgorithm replied:
The last review suggested removing it. I tried running this with SQLAlchemy and it seemed fine.

s3rius commented on fastapi_template/template/{{cookiecutter.project_name}}/deploy/docker-compose.yml:
```diff
@@ -56,9 +56,13 @@ services:
       {{cookiecutter.project_name | upper}}_DB_USER: {{cookiecutter.project_name}}
       {{cookiecutter.project_name | upper}}_DB_PASS: {{cookiecutter.project_name}}
+      {%- if cookiecutter.db_info.name == "mongodb" %}
+      {{cookiecutter.project_name | upper}}_DB_BASE: admin
```
Can we name the base with the name of the project as for other DBs?

usefulalgorithm replied:
That would require populating another config file called `mongo-init.js` and mounting it into the container. See https://stackoverflow.com/a/54064268 for more info.

Do you think it's better to include this in this PR, or I open another one to implement this? Either one works for me.

Thanks!

s3rius replied:
I guess separate PR is fine. Let's not think about it now. Because as I think it's quite complex to set up custom database name.

s3rius commented on fastapi_template/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/web/api/dummy/schema.py:
```diff
@@ -13,9 +17,26 @@ class DummyModelDTO(BaseModel):
     name: str
 
+    {%- if cookiecutter.db_info.name == "mongodb" %}
+    @field_validator("id", mode="before")
```
This thing won't work for old pydanticV1. 

You can check which pydantic version is currently being used by using `{%- if cookiecutter.pydanticv1 == "True" %} ...`

s3rius replied:
Also, I'm wondering whether we need that validation in the end? Because as I think, pydantic will try to cast the value to string.

usefulalgorithm replied:
> This thing won't work for old pydanticV1.

Good catch, I'll fix it.

>  I'm wondering whether we need that validation in the end?

We do need it, Pydantic models cannot store ObjectId unless we have
```
model_config = ConfigDict(extra="allow")
```

IMO allowing whatever third party class to be included in the model class seems a bit overkill, using a validator seems like a more reasonable way of doing it.

s3rius replied:
Agree. Extra seems like a possible bug source in the future. Because they might change the type and handlers will no longer be able to return it in responses.

s3rius commented on fastapi_template/template/{{cookiecutter.project_name}}/{{cookiecutter.project_name}}/web/lifetime.py:
```diff
@@ -120,6 +120,19 @@ def _setup_db(app: FastAPI) -> None:  # pragma: no cover
+from motor.motor_asyncio import AsyncIOMotorClient
+from {{cookiecutter.project_name}}.db.models import load_all_models
+async def _setup_db(app: FastAPI) -> None:
+    client = AsyncIOMotorClient(settings.db_url.human_repr())
```
```suggestion
    client = AsyncIOMotorClient(str(settings.db_url))
```

s3rius reviewed (changes requested):
Sorry. I couldn't find time to review. It takes a lot of time to review every request to this project. 

But thank you very much for implementing it. I know how much effort it takes to implement something like this.

s3rius commented:
Okay. I don't know what is happening, but now mypy seems to complain aboutsome types. Can you fix it? 

Just add an `[Any]` to every `AsyncConnectionPool` like `AsyncConnectionPool[Any]`. Don't forget to add any in imports if it doesn't exists yet in the file.

usefulalgorithm commented:
Sorry I missed a `AsyncConnectionPool`... can you trigger the test again?  Thanks

usefulalgorithm commented:
I don't get it...

s3rius commented:
@usefulalgorithm, that's fine. Happens sometimes. Will rerun tests.

s3rius commented:
I will try to fix the problem on develop branch. Thanks for your outstanding contribution.
